File: ex109.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        
        if nums == []:
            return None  
        
        mid = len(nums) // 2
        root = TreeNode(nums[mid])
        
        root.left = self.sortedArrayToBST(nums[ : mid])
        root.right = self.sortedArrayToBST(nums[mid + 1 : ])
        
        return root
    
    
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        
        nums = []
        
        #Create nums array
        while head != None:
            nums.append(head.val)
            head = head.next
            
        root = self.sortedArrayToBST(nums)
        
        return root
        

File: code/ex109.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    vector<int> list;
public:
    TreeNode* sortedListToBST(ListNode* head) {
        while (head != NULL) {
            list.push_back(head->val);
            head = head->next;
        }
        return build(0, list.size() - 1);
    }
    
private:
    TreeNode* build(int lo, int hi) {
        if (lo > hi)    return NULL;
        int mid = lo + (hi - lo) / 2;
        TreeNode* root = new TreeNode(list[mid]);
        root->left = build(lo, mid - 1);
        root->right = build(mid + 1, hi);
        return root;
    }
};